payroll/strategies/strategy_of_payments_interface.py

```diff
@@ -7,10 +7,53 @@ class StrategyOfPaymentInterface(object,  metaclass=abc.ABCMeta):
     def accept_payroll(cls, payroll, user, **kwargs):
         pass
 
+    @classmethod
+    def reject_payroll(cls, payroll, user, **kwargs):
+        from payroll.models import PayrollStatus
+        cls.change_status_of_payroll(payroll, PayrollStatus.REJECTED, user)
+        cls._remove_benefits_from_rejected_payroll(payroll)
+
     @classmethod
     def acknowledge_of_reponse_view(cls, payroll, response_from_gateway, user, rejected_bills):
         pass
 
     @classmethod
     def reconcile_payroll(cls, payroll, user):
         pass
+
+    @classmethod
+    def change_status_of_payroll(cls, payroll, status, user):
+        payroll.status = status
+        payroll.save(username=user.login_name)
+
+    @classmethod
+    def _remove_benefits_from_rejected_payroll(cls, payroll):
+        from payroll.models import (
+            BenefitAttachment,
+            BenefitConsumption,
+            PayrollBenefitConsumption
+        )
+        from invoice.models import (
+            Bill,
+            BillItem
+        )
+        benefits = list(BenefitConsumption.objects.filter(
+            payrollbenefitconsumption__payroll__id=payroll.id,
+            is_deleted=False
+        ).values_list('id', flat=True).distinct())
+        # remove from db all related fields to payroll, no business need to keep them in db
+        bill_items = list(
+            BillItem.objects.filter(
+                bill__benefitattachment__benefit_id__in=benefits
+            ).values_list('id', flat=True).distinct()
+        )
+        bills = list(
+            Bill.objects.filter(
+                benefitattachment__benefit_id__in=benefits
+            ).values_list('id', flat=True).distinct()
+        )
+        BenefitAttachment.objects.filter(benefit_id__in=benefits).delete()
+        BillItem.objects.filter(id__in=bill_items).delete()
+        Bill.objects.filter(id__in=bills).delete()
+        PayrollBenefitConsumption.objects.filter(payroll_id=payroll.id).delete()
+        BenefitConsumption.objects.filter(id__in=benefits).delete()
```

Comment:
I think we can make this part more efficient, espacially by removing the unnecessary casing to the list. 
Perhaps something like 
```python
BenefitAttachment.objects.filter(
    benefit_id__in=BenefitConsumption.objects.filter(
        payrollbenefitconsumption__payroll=payroll,
        is_deleted=False
    )
).delete()

BillItem.objects.filter(
    bill__benefitattachment__benefit__in=BenefitConsumption.objects.filter(
        payrollbenefitconsumption__payroll=payroll,
        is_deleted=False
    )
).delete()

Bill.objects.filter(
    benefitattachment__benefit__in=BenefitConsumption.objects.filter(
        payrollbenefitconsumption__payroll=payroll,
        is_deleted=False
    )
).delete()


PayrollBenefitConsumption.objects.filter(payroll=payroll).delete()

BenefitConsumption.objects.filter(
    payrollbenefitconsumption__payroll=payroll,
    is_deleted=False
).delete()
```
would do the trick.